client.go

```diff
@@ -918,11 +918,28 @@ func (client *client) tryRefreshMetadata(topics []string, attemptsRemaining int,
 				Logger.Println("client is not authorized to access this topic. The topics were: ", topics)
 				return err
 			}
+
+			if errors.Is(err, ErrUnsupportedSASLMechanism) {
+				Logger.Println("requested SASL mechanism is not supported by the broker")
+				return err
+			}
+
 			// else remove that broker and try again
 			Logger.Printf("client/metadata got error from broker %d while fetching metadata: %v\n", broker.ID(), err)
 			_ = broker.Close()
 			client.deregisterBroker(broker)
 		} else {
+			if errors.Is(err, ErrSASLHandshakeReadEOF) ||
+				errors.Is(err, ErrSASLHandshakeSendEOF) ||
+				errors.Is(err, ErrFetchMetadataEOF) ||
+				errors.Is(err, ErrBadTLSHandshake) {
+				// These errors are typically unrecoverable connection /
+				// authentication failures, so we return them
+				// directly here to avoid falling back on the less useful
+				// "client has run out of brokers" error after retrying.
+				return err
+			}
+
```

Comment:
Having looked at the test failures, and thought about this PR some more — do we still want to keep this "treat as unrecoverable" case here? Theoretically our client may have just been unluckly and it's connection with a broker happened to be disconnected (e.g., due to a restart) whilst it was refreshing metadata from that one — we don't want to hard fail there do we?

Now that Sarama will wrap all errors are return them (rather than just a "run out of brokers" message) perhaps the per-broker specific messages will be sufficient to help with diagnosing any issue?

Reply:
Yes so https://github.com/Shopify/sarama/pull/2131 made that available in wrapped errors